File: multiplayer/server/vector2d.go
package main

import "math"

type Vector2D struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

func NewVector2D(x, y float64) Vector2D {
	return Vector2D{X: x, Y: y}
}

func (v Vector2D) Add(other Vector2D) Vector2D {
	return NewVector2D(v.X+other.X, v.Y+other.Y)
}

func (v Vector2D) Multiply(scalar float64) Vector2D {
	return NewVector2D(v.X*scalar, v.Y*scalar)
}

func (v Vector2D) Magnitude() float64 {
	return math.Sqrt(v.X*v.X + v.Y*v.Y)
}

func (v Vector2D) Normalize() Vector2D {
	mag := v.Magnitude()
	if mag > 0 {
		return NewVector2D(v.X/mag, v.Y/mag)
	}
	return NewVector2D(0, 0)
}

File: multiplayer/client/src/vector2d.ts
// Your existing Vector2D class
export class Vector2D {
  constructor(public x: number, public y: number) {}

  add(other: Vector2D): Vector2D {
    return new Vector2D(this.x + other.x, this.y + other.y);
  }

  multiply(scalar: number): Vector2D {
    return new Vector2D(this.x * scalar, this.y * scalar);
  }

  magnitude(): number {
    return Math.sqrt(this.x * this.x + this.y * this.y);
  }

  normalize(): Vector2D {
    const mag = this.magnitude();
    return mag > 0
      ? new Vector2D(this.x / mag, this.y / mag)
      : new Vector2D(0, 0);
  }
}
